rcl/src/rcl/subscription.c

```diff
@@ -193,6 +209,238 @@ rcl_subscription_get_default_options()
+      option->rmw_subscription_options.content_filter_options, allocator);
+    if (RCUTILS_RET_OK != ret) {
+      RCUTILS_SAFE_FWRITE_TO_STDERR("Failed to fini content filter options.\n");
+      return rcl_convert_rmw_ret_to_rcl_ret(ret);
+    }
+    allocator->deallocate(
+      option->rmw_subscription_options.content_filter_options, allocator->state);
+    option->rmw_subscription_options.content_filter_options = NULL;
+  }
+  return RCL_RET_OK;
+}
+
+rcl_ret_t
+rcl_subscription_options_set_content_filter_options(
+  const char * filter_expression,
+  size_t expression_parameters_argc,
+  const char * expression_parameter_argv[],
+  rcl_subscription_options_t * options)
+{
+  RCL_CHECK_ARGUMENT_FOR_NULL(filter_expression, RCL_RET_INVALID_ARGUMENT);
+  if (expression_parameters_argc > 100) {
+    RCL_SET_ERROR_MSG("The maximum of expression parameters argument number is 100");
+    return RCL_RET_INVALID_ARGUMENT;
+  }
+  RCL_CHECK_ARGUMENT_FOR_NULL(options, RCL_RET_INVALID_ARGUMENT);
+  const rcl_allocator_t * allocator = &options->allocator;
+  RCL_CHECK_ALLOCATOR_WITH_MSG(allocator, "invalid allocator", return RCL_RET_INVALID_ARGUMENT);
+
+  rcl_ret_t ret;
+  rmw_subscription_content_filter_options_t * content_filter_options =
+    allocator->allocate(
+    sizeof(rmw_subscription_content_filter_options_t), allocator->state);
+  if (!content_filter_options) {
+    RCL_SET_ERROR_MSG("allocating memory failed");
+    return RCL_RET_BAD_ALLOC;
+  }
+
+  *content_filter_options = rmw_get_zero_initialized_content_filter_options();
+
+  rmw_ret_t rmw_ret = rmw_subscription_content_filter_options_set(
+    filter_expression,
+    expression_parameters_argc,
+    expression_parameter_argv,
+    allocator,
+    content_filter_options
+  );
+
+  if (rmw_ret != RMW_RET_OK) {
+    ret = rcl_convert_rmw_ret_to_rcl_ret(rmw_ret);
+    goto failed;
+  }
+
+  if (options->rmw_subscription_options.content_filter_options) {
+    rmw_ret = rmw_subscription_content_filter_options_fini(
+      options->rmw_subscription_options.content_filter_options,
+      allocator
+    );
+
+    if (rmw_ret != RMW_RET_OK) {
+      return rcl_convert_rmw_ret_to_rcl_ret(rmw_ret);
+    }
+
+    allocator->deallocate(
+      options->rmw_subscription_options.content_filter_options, allocator->state);
+    options->rmw_subscription_options.content_filter_options = NULL;
+  }
+  options->rmw_subscription_options.content_filter_options = content_filter_options;
+
+  return RCL_RET_OK;
+
+failed:
+  allocator->deallocate(content_filter_options, allocator->state);
+  return ret;
+}
+
+rcl_subscription_content_filter_options_t
+rcl_get_zero_initialized_subscription_content_filter_options()
+{
+  return (const rcl_subscription_content_filter_options_t) {
+           .allocator = rcl_get_default_allocator(),
+           .rmw_subscription_content_filter_options =
+             rmw_get_zero_initialized_content_filter_options()
+  };  // NOLINT(readability/braces): false positive
```

Comment:
indentantion looks weird here

```suggestion
  return (const rcl_subscription_content_filter_options_t) {
      .allocator = rcl_get_default_allocator(),
      .rmw_subscription_content_filter_options =
        rmw_get_zero_initialized_content_filter_options()
  };  // NOLINT(readability/braces): false positive
```

Reply:
The checking will be failed in `ament_uncrustify`.
```
@@ -301,3 +301,3 @@
-      .allocator = rcl_get_default_allocator(),
-      .rmw_subscription_content_filter_options =
-        rmw_get_zero_initialized_content_filter_options()
+           .allocator = rcl_get_default_allocator(),
+           .rmw_subscription_content_filter_options =
+             rmw_get_zero_initialized_content_filter_options()
```